packages/react/src/components/ContextMenu/ContextMenu.tsx

```diff
@@ -0,0 +1,234 @@
+  }
+
+  closeAndFocus(e: Event, which: 'next' | 'previous') {
+    const renderCallback = () => focusNearest(this.triggerRef.current, which)
+    this.setState(
+      {
+        open: false,
+        autoFocus: false,
+      },
+      renderCallback,
+    )
+    e.preventDefault()
+  }
+
+  openAndFocus(e: Event, which: 'first' | 'last') {
+    const renderCallback = () => focusMenuItem(this.menuRef.current, which)
+    this.setState(
+      {
+        open: true,
+        autoFocus: false, // focused by renderCallback
+      },
+      renderCallback,
+    )
+    e.preventDefault()
+  }
+
+  handleOpenChange = (e, { open }) => {
+    _.invoke(this.props, 'onOpenChange', e, { ...this.props, ...{ open } })
+    this.setState(() => ({
+      open,
+      autoFocus: true,
+    }))
+  }
+
+  handleMenuItemClick = (predefinedProps?: MenuItemProps) => (
+    e: React.SyntheticEvent,
+    itemProps: MenuItemProps,
+  ) => {
+    _.invoke(predefinedProps, 'onClick', e, itemProps)
+    if (!predefinedProps || !predefinedProps.menu) {
+      // do not close if clicked on item with submenu
+      this.setState({ open: false, autoFocus: false })
+    }
+  }
+
+  handleMenuItemOverrides = (menuItemAccessibilityAttributes: AccessibilityAttributes) =>
+    _.map(_.get(this.props.menu, 'items'), (item: ShorthandValue<MenuItemProps>) =>
+      typeof item === 'object'
+        ? {
+            ...item,
+            onClick: this.handleMenuItemClick(item as MenuItemProps),
+            ...menuItemAccessibilityAttributes,
+          }
+        : {
+            content: item,
+            key: item,
+            onClick: this.handleMenuItemClick(),
+            ...menuItemAccessibilityAttributes,
+          },
+    )
+
+  renderComponent({
+    ElementType,
+    classes,
+    unhandledProps,
+    accessibility,
+  }: RenderResultConfig<ContextMenuProps>): React.ReactNode {
+    const { menu, ...popupProps } = this.props
+    const content =
+      menu &&
+      Menu.create(menu, {
+        defaultProps: {
+          ...accessibility.attributes.menu,
+          vertical: true,
+        },
+        overrideProps: {
+          items: this.handleMenuItemOverrides(accessibility.attributes.menuItem),
+        },
+      })
+
+    return (
+      <ElementType
+        className={classes.root}
+        {...accessibility.attributes.root}
+        {...unhandledProps}
+        {...applyAccessibilityKeyHandlers(accessibility.keyHandlers.root, unhandledProps)}
+        style={{ boxSizing: 'border-box', display: 'inline-block' }}
+      >
+        <Ref innerRef={this.triggerRef}>
+          <Popup
+            {...popupProps}
+            accessibility={() => accessibility}
+            open={this.state.open}
+            onOpenChange={this.handleOpenChange}
+            inline
+            autoFocus
+            content={{
+              variables: { padding: '', borderSize: '0px' }, // TODO: this should probably be in the context menu variables?
```

Comment:
I don't like this, hoping to find a better solution - for example make these two a variable of the ContextMenu itself?

Reply:
override styles, create them for menu slot, pass them down to popup and popup content